File: 2024/day11/day11.py
def blink(stone):
    if stone == '0':
        return '1'
    elif len(stone)%2 == 0:
        left = str(int(stone[:len(stone)//2]))
        right = str(int(stone[len(stone)//2:]))
        return [left, right]
    else:
        return [str(2024*int(stone))]

with open("day11.txt", "r") as f:
    stones = (f.read().split("\n")[0]).split(" ")

stone_dict = {}
for stone in stones:
    if stone in stone_dict.keys():
        stone_dict[stone]+=1
    else:
        stone_dict[stone] = 1
for i in range(75):
    new_stone_dict = {}
    for stone in stone_dict.keys():
        for new_stone in blink(stone):
            if new_stone in new_stone_dict.keys():
                new_stone_dict[new_stone] += stone_dict[stone]
            else:
                new_stone_dict[new_stone] = stone_dict[stone]
    stone_dict = new_stone_dict
print(sum([stone_dict[x] for x in stone_dict.keys()]))

File: 2024/day11/day11.hs
import Data.List
import qualified Data.Map as Map
import System.IO

blink_many :: Map.Map String Int -> String -> Int -> Map.Map String Int
blink_many scores label count = 
    if label == "0" then (Map.insertWith (+) "1" count scores)
    else
        let n = length label
        in
            if (n`mod` 2) == 0 then
                let
                    left = show((read::String->Int)(take (n`div`2) label))
                    right = show((read::String->Int)(drop (n`div`2) label))
                in
                    (Map.insertWith (+) right count (Map.insertWith (+) left count scores))
            else (Map.insertWith (+) (show (2024*read label)) count scores)
main = do
    handle <- openFile "day11.txt" ReadMode
    contents <- hGetContents handle
    let
        numbers = words contents
        initial = [(x, 1)|x<-numbers] --there are no repetitions; I don't want to write a whole thing to count
        final_stones = (iterate (Map.foldlWithKey blink_many Map.empty) (Map.fromList initial))!!75
    print(sum (Map.elems final_stones))
